### PR TITLE
Handle checking smartgroups when generating list of groups to unsubscribe

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
The unsubscribe form doesn't check if you've already unsubscribed from a smartgroup so it looks like it doesn't work when you do so because it always shows the smartgroups as available for unsubscribe.  That's not a very good user experience, especially as for many users this will be their last interaction with the organisation.

Before
----------------------------------------
Always shows this first page if there are unsubscribable smartgroups even if the user has already unsubscribed from those groups:
image

After
----------------------------------------
Only shows the first page if there are groups/smartgroups that the user is still a member of. Otherwise just displays a message that they've already unsubscribed.

Technical Details
----------------------------------------
"Borrows" from CiviRules and elsewhere. We check (and refresh if necessary) the groups to make sure we have an up to date group contact cache. Then we run a query that checks both `civicrm_group_contact` and `civicrm_group_contact_cache`.

Comments
----------------------------------------

### CODE REVIEW

**Comment by seamuslee001** [comment]
So thinking about this from a performance POV, smart groups aren't always that performant, that having being said we are limited to a very select number of groups so I think we would be ok, but might be worth getting someone like @johntwyman to look at this one

**Comment by mattwire** [comment]
@seamuslee001 I was a bit concerned about any performance impact but I don't think it should be that bad because:
- smartgroup is only rebuilt if cache time has expired.
- rebuild will only be triggered if someone clicks the unsubscribe link
- if multiple users click the unsubscribe link at a similar time it's likely they'll be subscribed to the same groups so should only trigger one or two rebuilds of expired smartgroups.
- can only happen after a mailing has been sent.

**Comment by mlutfy** [comment]
There are days where I wish we did not allow Smart Groups to be used as mailing groups. People abuse it and it gives really odd results like this (I was testing an unsubscribe from the "Campfire" mailings):

unsubscribe-2021-09-06_14-16

(ignore the meme, I was confused at the form re-appearing after having unsubscribed, other issue)

That said - if people do abuse of smartgroups, then I think merging this is the right thing to do.

File: CRM/Mailing/Event/BAO/Unsubscribe.php [hunk 1]
```diff
@@ -226,23 +226,33 @@ public static function unsub_from_mailing($job_id, $queue_id, $hash, $return = F
...
                         grp.title as title,
                         grp.frontend_title as frontend_title,
                         grp.frontend_description as frontend_description,
-                        grp.description as description
+                        grp.description as description,
+                        grp.saved_search_id as saved_search_id
             FROM        civicrm_group grp
             LEFT JOIN   civicrm_group_contact gc
                 ON      gc.group_id = grp.id
+                LEFT JOIN civicrm_group_contact_cache gcc
```
Comment (kurund) [inline CRM/Mailing/Event/BAO/Unsubscribe.php:243]:
```suggestion
            LEFT JOIN     civicrm_group_contact_cache gcc
```

File: CRM/Mailing/Event/BAO/Unsubscribe.php [hunk 2]
```diff
@@ -226,23 +226,33 @@ public static function unsub_from_mailing($job_id, $queue_id, $hash, $return = F
...
             WHERE       grp.is_hidden = 0
                         $groupIdClause
-                AND     (grp.saved_search_id is not null
-                            OR  (gc.contact_id = $contact_id
+                AND     ((grp.saved_search_id is not null AND gcc.contact_id = %1)
+                            OR  (gc.contact_id = %1
                                 AND gc.status = 'Added')
                             $baseGroupClause
-                        )");
+                        ) GROUP BY grp.id;";
```
Comment (kurund) [inline CRM/Mailing/Event/BAO/Unsubscribe.php:251]:
```suggestion
                        ) GROUP BY grp.id";
```

**Review by kurund — CHANGES_REQUESTED** [review]
@mattwire 
Just added some minor formatting changes and removed extra ;

**Comment by kurund** [comment]
@mattwire @seamuslee001 

This fix make sense to me.  I also think it won't have major impact on performance due to smart group rebuilding etc for the reasons mentioned above.

I have tested the patch and it is working as expected.

**Comment by mattwire** [comment]
Thanks @kurund I've committed your suggestions and squashed commits

**Comment by mlutfy** [comment]
@seamuslee001 @johntwyman Any final thoughts? Need more time?

**Comment by johntwyman** [comment]
Thanks for the invitation to give this some thought. On balance I think improving the user experience trumps the performance cost of recalculating smart group membership. I am somewhat concerned about that for an installation like ours but the reality is we're the creators of our own misfortune in many regards. ;-)

**Comment by mattwire** [comment]
@kurund For me yes!

**Comment by tommybobo** [comment]
I am leaving this here for anyone else stuck in a pickle. 

So this change just crashed site yesterday and the day before, It is yes a mess of their own making but also I think a key performance issue was over looked. I have noticed that a rise in enterprise level mail servers having bots check and scan links in emails. When a large list is emailed with a large number of enterprise emails, those bots quickly check every link including the unsubscribe link. So we found a site basically crippled by trying to send out on an email while this unsubscribe look up was being run 20ish times a minute. And stalling out because the links were being clicked nearly simultaneously and thus seeming to regenerate the cache simultaneously. Also while the link bots were loading pages, logging clicks, and the next batch of emails were being sent.

We are looking into ways to circumvent this pile up. In the meantime we are increasing the cache timeout, using only Opt Out of all, and throttling outbound emails significantly.

**Comment by eileenmcnaughton** [comment]
Regression logged against this https://lab.civicrm.org/dev/core/-/issues/3031

Performance impact was discussed in this ticket earlier but it was said that 

"smartgroup is only rebuilt if cache time has expired."

It is more correct to say 'smartgroups are rebuilt for ALL groups whose cache time has expired' - in the case of AUG I would think that would be a couple of thousand groups - which is why we previously removed most loadAll calls from the code

**Pull request merged**